java/rocksjni/flink_compactionfilterjni.cc

```diff
@@ -0,0 +1,175 @@
+    CheckAndRethrowException(env);
+    env->DeleteLocalRef(jlist);
+    releaseJniEnv(attached_thread);
+    return static_cast<std::size_t>(next_offset);
+  };
+
+private:
+  inline void CheckAndRethrowException(JNIEnv* env) const {
+    if (env->ExceptionCheck()) {
+      env->ExceptionDescribe();
+      env->Throw(env->ExceptionOccurred());
+    }
+  }
+
+  jmethodID m_jnext_unexpired_offset_methodid;
+};
+
+class JavaListElementIterFactory : public rocksdb::flink::FlinkCompactionFilter::ListElementIterFactory, rocksdb::JniCallback {
+public:
+    JavaListElementIterFactory(JNIEnv* env, jobject jlist_iter_factory) : JniCallback(env, jlist_iter_factory) {
+        jclass jclazz = rocksdb::JavaClass::getJClass(env, "org/rocksdb/FlinkCompactionFilter$ListElementIterFactory");
+        if(jclazz == nullptr) {
+            // exception occurred accessing class
+            return;
+        }
+        m_jcreate_iter_methodid = env->GetMethodID(jclazz,
+                "createListElementIter", "()Lorg/rocksdb/FlinkCompactionFilter$ListElementIter;");
+        assert(m_jcreate_iter_methodid != nullptr);
+    }
+
+    FlinkCompactionFilter::ListElementIter* CreateListElementIter(std::shared_ptr<rocksdb::Logger> /*logger*/) const override {
+        jboolean attached_thread = JNI_FALSE;
+        JNIEnv* env = getJniEnv(&attached_thread);
+        auto jlist_iter = env->CallObjectMethod(m_jcallback_obj, m_jcreate_iter_methodid);
+        auto list_iter = new JavaListElementIter(env, jlist_iter);
+        CheckAndRethrowException(env);
+        releaseJniEnv(attached_thread);
+        return list_iter;
+    };
+
+private:
+    inline void CheckAndRethrowException(JNIEnv* env) const {
+        if (env->ExceptionCheck()) {
+            env->ExceptionDescribe();
+            env->Throw(env->ExceptionOccurred());
+        }
+    }
+
+    jmethodID m_jcreate_iter_methodid;
+};
+
+static FlinkCompactionFilter::ListElementIterFactory* createListElementIterFactory(
+        JNIEnv* env, jint ji_list_elem_len, jobject jlist_iter_factory) {
+  FlinkCompactionFilter::ListElementIterFactory* list_iter_factory = nullptr;
+  if (ji_list_elem_len > 0) {
+    auto fixed_size = static_cast<std::size_t>(ji_list_elem_len);
+    list_iter_factory = new FlinkCompactionFilter::FixedListElementIterFactory(fixed_size, static_cast<std::size_t>(0));
+  } else if (jlist_iter_factory != nullptr) {
+    list_iter_factory = new JavaListElementIterFactory(env, jlist_iter_factory);
+  }
+  return list_iter_factory;
+}
+
+/*
+ * Class:     org_rocksdb_FlinkCompactionFilter
+ * Method:    createNewFlinkCompactionFilterConfigHolder
+ * Signature: ()J
+ */
+jlong Java_org_rocksdb_FlinkCompactionFilter_createNewFlinkCompactionFilterConfigHolder(
+        JNIEnv* /* env */, jclass /* jcls */) {
+    using namespace rocksdb::flink;
+    return reinterpret_cast<jlong>(
+            new std::shared_ptr<FlinkCompactionFilter::ConfigHolder>(new FlinkCompactionFilter::ConfigHolder()));
```

Comment:
1. when you combine create a `shared_ptr` with new allocation, you can use make `make_shared` to combine them into one allocation.
2. What is the value of wrapping into a `shared_ptr` here, at all - it doesn't seem like a shared resource and you explictly delete in dispose?